instrumentation-core/jmx-metrics/README.md

```diff
@@ -0,0 +1,110 @@
+# JMX Metric Gatherer
+
+This utility provides an easy framework for gathering and reporting metrics based on queried
+MBeans from a JMX server.  It loads a custom Groovy script and establishes a helpful, bound `otel`
+object with methods for obtaining MBeans and constructing synchronous OpenTelemetry instruments:
+
+### Usage
+
+```bash
+$ java -D<otel.jmx.metrics.property=value> -jar opentelemetry-jmx-metrics-<version>-all.jar [-config ./optional_config.properties]
+```
+
+##### `optional_config.properties` example
+
+```properties
+otel.jmx.metrics.service.url = service:jmx:rmi:///jndi/rmi://<my-jmx-host>:<my-jmx-port>/jmxrmi
+otel.jmx.metrics.groovy.script = /opt/script.groovy
+otel.jmx.metrics.interval.seconds = 5
+otel.jmx.metrics.exporter.type = otlp
+otel.jmx.metrics.exporter.endpoint = my-opentelemetry-collector:55680
+otel.jmx.metrics.username = my-username
+otel.jmx.metrics.password = my-password
+```
+
+##### `script.groovy` example
+
+```groovy
+import io.opentelemetry.common.Labels
+
+def loadMatches = otel.queryJmx("org.apache.cassandra.metrics:type=Storage,name=Load")
+def load = loadMatches.first()
+
+def lvr = otel.longValueRecorder(
```

Comment:
I think a lot of the industry has switched from inline scripts to standard static configs, often generating them in something like terraform, mainly because of security dangers of inline scripts. Do you have any thoughts on the scripting vs config approach?

Exporting metrics for arbitrary management beans seems like an awesome feature I'd want to enable, but I would be quite wary in a production environment due to the security concern (I have no problem generating configs with another tool though)

Reply:
This is a good point, and I make no pretense to speak authoritatively but only from personal preference.  Config based jmx solutions like jmxtrans and collectd's [GenericJMX](https://collectd.org/wiki/index.php/Plugin:GenericJMX) are powerful but have few or no capabilities for aggregations or transformations, and are generally unexpressive.  They are also isolated and without ways of producing custom metrics informed by mbean state.  Manual instrumentation seems preferable in light of this, though improving the helper utility to support a more declarative style of instrumentation could be useful.

As far as security concerns they seem to be the same as a lot of dynamic configuration management utilities.  Guidelines for restricting permissions via an access file and configuring the security manager could be helpful to reduce the potential for destructive purposes.